includes/Core/Authentication/Clients/OAuth_Client.php

```diff
@@ -404,10 +404,54 @@ public function get_required_scopes() {
 	 * @since 1.0.0
 	 * @see https://developers.google.com/identity/protocols/googlescopes
 	 *
-	 * @return array List of Google OAuth scopes.
+	 * @return string[] List of Google OAuth scopes.
 	 */
 	public function get_granted_scopes() {
-		return array_values( (array) $this->user_options->get( self::OPTION_AUTH_SCOPES ) );
+		$base_scopes  = $this->user_options->get( self::OPTION_AUTH_SCOPES ) ?: array();
+		$extra_scopes = $this->get_granted_additional_scopes();
+
+		return array_unique(
+			array_merge( $base_scopes, $extra_scopes )
+		);
+	}
+
+	/**
+	 * Gets the list of currently granted additional Google OAuth scopes for the current user.
+	 *
+	 * @since n.e.x.t
+	 * @see https://developers.google.com/identity/protocols/googlescopes
+	 *
+	 * @return string[] List of Google OAuth scopes.
+	 */
+	public function get_granted_additional_scopes() {
+		return array_values( $this->user_options->get( self::OPTION_ADDITIONAL_AUTH_SCOPES ) ?: array() );
+	}
+
+	/**
+	 * Checks if new scopes are required that are not yet granted for the current user.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool true if any required scopes are not satisfied, otherwise false.
+	 */
+	public function requires_additional_scopes() {
+		if ( ! $this->get_access_token() ) {
+			return false;
+		}
+
+		return ! $this->has_sufficient_scopes( $this->get_required_scopes() );
+	}
```

Comment:
Can we rename this to `needs_reauthentication()`? This is closer to how it is in JS and how it was before. The current name is slightly confusing because "additional scopes" has a special meaning for us (scopes that are not the base required scopes but the user has granted them).

Reply:
I thought about that and "reauthentication" is a bit vague.  which is that there are required scopes which are not satisfied. I hear you about the inconsistency of terms though. Maybe `has_insufficient_required_scopes` ? I don't mean to bike shed on this but just giving it a bit more thought now that it's becoming public.

I'll rename this for now in the interest of time, but let me know if you have any other ideas.

Reply:
I think the `needs_reauthentication` is fairly clear for this boolean value, given that it's a boolean I would argue there isn't even a need to know the "internals" on why reauthentication is needed - the other thing is we've been using this terminology throughout our codebase, I don't think it's worth changing.